pkg/engine/resolve/dataloader.go

```diff
@@ -0,0 +1,119 @@
+package resolve
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/buger/jsonparser"
+
+	"github.com/jensneuse/graphql-go-tools/pkg/fastbuffer"
+)
+
+type DataLoader struct {
+	totalNum           int
+	inputs             [][]byte
+	inputPosToBufPairs map[int]*BufPair
+	fetch              *BatchFetch
+
+	batch *batch
+
+	mu *sync.Mutex
+}
+
+type batch struct {
+	err  error
+	done chan struct{}
+}
+
+func (d *DataLoader) Load(ctx *Context, input []byte, bufPair *BufPair) (err error) {
```

Comment:
This func is very smart.
At the same time, we're creating one goroutine for each of the N-fetches, all of them blocking until the batch goroutine finishes.
Maybe we find a way to collect all N-fetches within the main goroutine and also batch-fetch without an additional goroutine. I don't see why we need all these goroutines when we're doing work in only one of them.

Reply:
> This func is very smart.
I have used this [pkg](https://github.com/vektah/dataloaden) with [gqlgen](https://gqlgen.com/) and tried to make something similar (my Dataloader has similar API and implementation).

> At the same time, we're creating one goroutine for each of the N-fetches, all of them blocking until the batch goroutine finishes.
The solution is based on asynchronous resolve of array (**resolveArrayAsynchronous** method). The method (**resolveArrayAsynchronous**) is responsible for creating goroutine.  We can utilize Worker Pool for preventing unnecessarily large number of goroutines.

Let me give you an example:
Query 
# Write your query or mutation here
query {
  topProducts {
    reviews {
      author {
        someProperyReqiresFetchingOfAuthor
      }
    }
  }
}

![Screenshot from 2021-06-07 15-00-34](https://user-images.githubusercontent.com/12982070/121014757-fdb41900-c7a2-11eb-8730-b2b22a0100cb.png)
At **point C** we must gather all **authors**, deduplicate input (do not repeat the same **author** in **request**) and make single batch request.  It's not obvious how to get to this point if  we do not resolve **Array of Products** (Point A)  and **Array of reviews** (Point B) concurrently.
Synchronous resolve implies that  at **Point A** we  resolve whole branch for **Product A** and only after that we start resolving branch for **Product B**,  am I right?

Reply:
I think you're missing something, but maybe my thinking is wrong. Let's discuss..

So first, we resolve topProducts which returns an array of products.
At this point, we know the value of len(topProducts), which will be important later on.

Next, we start resolving Product A synchronously.
We'll enter the reviews Node, which the query planner market as batch.
We also know the size of the batch = len(topProducts) from above.
This means, we can add all required review fetches to the batch and then walk "backwards", one node up into the next Product Node.
This is possible because the query planner marked the Node as "batch".

Now, we enter the Product B Node and also add its reviews to the batch.
We know, the batch is not yet full so we walk back up and into Product C.

Same procedure for Product C.

Once the batch is full, we can actually do the fetch.
Now we un-batch the data and prepare it so that each of the datasources is capable of using it for further resolving.

Next, we're able to re-walk into Product A.
The fetch for the reviews is pre-resolved from the batch fetch.
That is, instead of doing a network fetch, we just fetch from the unbatched data and resolve Review A1, A2 & A3.

We see that we have to fetch Authors, which again are marked as batch.
We also know the size of the batch: the sum of all reviews (6)
So we add the first 3 Authors to the batch and walk upwards to the next root node.

Same procedure as before, just one level deeper, filling the author batch.

Then we resolve the author batch, un-batch and "pre load" all author batches.
Finally, we're able to re-walk into all Author Nodes and resolve them.

In total, this was 3 fetches (products + reviews + authors),
all within one single goroutine.
There was no concurrency needed.

Concurrency would come into play if there were child nodes without batching.
In that case, we'd asynchronously resolve each of them and then re-assemble the response.

What do you think about this approach?

Reply:
I agree, it's not ok to change resolving strategy only for gathering batches.
I have prepared [draft](https://github.com/chedom/graphql-go-tools/pull/2) with another solution. It's based on synchronous strategy.